### PR TITLE
Update: `props` option of `no-param-reassign` (fixes #1600)

### LINKED ISSUE #1600: Rule suggestion: Avoid modifying object references
In my opinion it is generally dangerous to change the input parameter of a function if it is not a primitive type. It could e.g. result in side effects that are hard to debug because the referenced object is changed in the function.

While I think we won't be able to detect every possible fault, I suggest creating a rule that at least catches the simple case:

```js
// These cases could be an error or a warning if the rule is enabled
function foo1 (bar) {
    bar.thisPropertyAlreadyExisted = somethingElse;
}

function foo2 (bar) {
    delete bar.thisPropertyAlreadyExisted;
}

// This is most likely to hard to implement
function foo3 (bar) {
    var temp = bar.thisPropertyAlreadyExisted;
    temp.deepProperty = somethingElse;
}
```

What do you think about this?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/7195788-rule-suggestion-avoid-modifying-object-references?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by mysticatea**

### CODE REVIEW

**Comment by mysticatea** [comment]
Oh, sorry, I forgot updating scheme.

**Comment by mysticatea** [comment]
I had not `npm update` for some time, my local test has been passing.

I modified `.\node_modules\eslint-tester\node_modules\eslint\lib\config-validator.js` in local for https://github.com/eslint/eslint-tester/issues/24 and ran tests. Then, I found several problems about schema, so I fixed those together.

File: docs/rules/no-param-reassign.md [hunk 1]
```diff
@@ -6,7 +6,21 @@ Assignment to variables declared as function parameters can be misleading and le
 
 This rule aims to prevent unintended behavior caused by overwriting function parameters.
 
-The following patterns are considered warnings:
+### Options
+
+This rule have an option.
+The option is an object which has a property `"props"`.
```
Comment (IanVS) [inline docs/rules/no-param-reassign.md]: I'd suggest:

```
This rule takes one option, an object, with a property `"props"`.
```

**Comment by nzakas** [comment]
Can you try rebasing?

**Comment by mysticatea** [comment]
Sure, I rebased.

**Comment by nzakas** [comment]
Lgtm

**Pull request merged**